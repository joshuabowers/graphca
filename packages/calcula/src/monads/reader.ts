export type Reader<E, T> = (env: E) => T

export const unit = <E, T>(t: T): Reader<E, T> => _ => t

export const bind = <E, A>(r: Reader<E, A>) => 
  <B>(fn: (a: A) => Reader<E, B>): Reader<E, B> =>
    (e: E) => fn(r(e))(e)

export const ask = <E>(env: E) => env

export const local = <E, T>(fn: (env: E) => E) =>
  (c: Reader<E, T>) =>
    (e: E) => c(fn(e))
