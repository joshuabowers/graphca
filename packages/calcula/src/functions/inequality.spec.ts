import { unit } from '../monads/writer'
import { 
  expectWriterTreeNode,
  realOps, complexOps, booleanOps, variableOps,
  equalsOps, notEqualsOps, lessThanOps, 
  greaterThanOps, lessThanEqualsOps, greaterThanEqualsOps
} from '../utility/expectations';
import { Clades, Genera, Species } from '../utility/tree';
import { real, complex, boolean } from '../primitives';
import { variable } from '../variable'
import {
  equals, notEquals, lessThan, greaterThan, lessThanEquals, greaterThanEquals, 
  $equals, $notEquals, $lessThan, $greaterThan, $lessThanEquals, $greaterThanEquals
} from "./inequality";

describe('$equals', () => {
  it('generates an Equality for a pair of TreeNode inputs', () => {
    expect(
      $equals(unit(variable('x').value), unit(variable('y').value))[0]
    ).toEqual({
      clade: Clades.binary, genus: Genera.inequalities, species: Species.equals,
      left: unit(variable('x').value), right: unit(variable('y').value)
    })
  })
})

describe('equals', () => {
  it('returns true for two equal reals', () => {
    expectWriterTreeNode(
      equals(real(1), real(1)),
      boolean(true)
    )(
      ...equalsOps(
        'real equality',
        realOps('1'),
        realOps('1'),
        booleanOps('true')
      )
    )
  })

  it('returns true for two equal complexes', () => {
    expectWriterTreeNode(
      equals(complex([1,1]), complex([1,1])),
      boolean(true)
    )(
      ...equalsOps(
        'complex equality',
        complexOps('1', '1'),
        complexOps('1', '1'),
        booleanOps('true')
      )
    )
  })

  it('returns true for two equal booleans', () => {
    expectWriterTreeNode(
      equals(boolean(false), boolean(false)),
      boolean(true)
    )(
      ...equalsOps(
        'boolean equality',
        booleanOps('false'),
        booleanOps('false'),
        booleanOps('true')
      )
    )
  })

  it('returns false for unequal things', () => {
    expectWriterTreeNode(
      equals(real(1), real(2)),
      boolean(false)
    )(
      ...equalsOps(
        'real equality',
        realOps('1'),
        realOps('2'),
        booleanOps('false')
      )
    )
  })

  it('returns an Equals for variable input', () => {
    expectWriterTreeNode(
      equals(variable('x'), variable('y')),
      $equals(variable('x'), variable('y'))[0]
    )(
      ...equalsOps(
        'created equality',
        variableOps('x'),
        variableOps('y'),
        []
      )
    )
  })
})

describe('$notEquals', () => {
  it('generates an Inequality for a pair of TreeNode inputs', () => {
    expect(
      $notEquals(unit(variable('x').value), unit(variable('y').value))[0]
    ).toEqual({
      clade: Clades.binary, genus: Genera.inequalities, species: Species.notEquals,
      left: unit(variable('x').value), right: unit(variable('y').value)
    })
  })
})

describe('notEquals', () => {
  it('returns false for two equal reals', () => {
    expectWriterTreeNode(
      notEquals(real(1), real(1)),
      boolean(false)
    )(
      ...notEqualsOps(
        'real inequality',
        realOps('1'),
        realOps('1'),
        booleanOps('false')
      )
    )
  })

  it('returns false for two equal complexes', () => {
    expectWriterTreeNode(
      notEquals(complex([1,1]), complex([1,1])),
      boolean(false)
    )(
      ...notEqualsOps(
        'complex inequality',
        complexOps('1', '1'),
        complexOps('1', '1'),
        booleanOps('false')
      )
    )
  })

  it('returns false for two equal booleans', () => {
    expectWriterTreeNode(
      notEquals(boolean(false), boolean(false)),
      boolean(false)
    )(
      ...notEqualsOps(
        'boolean inequality',
        booleanOps('false'),
        booleanOps('false'),
        booleanOps('false')
      )
    )
  })

  it('returns true for unequal things', () => {
    expectWriterTreeNode(
      notEquals(real(1), real(2)),
      boolean(true)
    )(
      ...notEqualsOps(
        'real inequality',
        realOps('1'),
        realOps('2'),
        booleanOps('true')
      )
    )
  })

  it('returns a NotEquals for variable input', () => {
    expectWriterTreeNode(
      notEquals(variable('x'), variable('y')),
      $notEquals(variable('x'), variable('y'))[0]
    )(
      ...notEqualsOps(
        'created inequality',
        variableOps('x'),
        variableOps('y'),
        []
      )
    )
  })
})

describe('$lessThan', () => {
  it('generates a LessThan for a pair of TreeNode inputs', () => {
    expect(
      $lessThan(unit(variable('x').value), unit(variable('y').value))[0]
    ).toEqual({
      clade: Clades.binary, genus: Genera.inequalities, species: Species.lessThan,
      left: unit(variable('x').value), right: unit(variable('y').value)
    })
  })
})

describe('lessThan', () => {
  it('returns true for two ordered reals', () => {
    expectWriterTreeNode(
      lessThan(real(1), real(2)),
      boolean(true)
    )(
      ...lessThanOps(
        'real less than',
        realOps('1'),
        realOps('2'),
        booleanOps('true')
      )
    )
  })

  it('returns false for two unordered reals', () => {
    expectWriterTreeNode(
      lessThan(real(2), real(1)),
      boolean(false)
    )(
      ...lessThanOps(
        'real less than',
        realOps('2'),
        realOps('1'),
        booleanOps('false')
      )
    )
  })

  it('returns false for two equal inputs', () => {
    expectWriterTreeNode(
      lessThan(real(1), real(1)),
      boolean(false)
    )(
      ...lessThanOps(
        'real less than',
        realOps('1'),
        realOps('1'),
        booleanOps('false')
      )
    )
  })

  it('returns true for two complexes ordered by length', () => {
    expectWriterTreeNode(
      lessThan(complex([1, 1]), complex([5, 5])),
      boolean(true)
    )(
      ...lessThanOps(
        'complex less than',
        complexOps('1', '1'),
        complexOps('5', '5'),
        booleanOps('true')
      )
    )
  })

  it('returns false for two complexes not ordered by length', () => {
    expectWriterTreeNode(
      lessThan(complex([5, 5]), complex([1, 1])),
      boolean(false)
    )(
      ...lessThanOps(
        'complex less than',
        complexOps('5', '5'),
        complexOps('1', '1'),
        booleanOps('false')
      )
    )
  })

  it('returns a LessThan for variable input', () => {
    expectWriterTreeNode(
      lessThan(variable('x'), variable('y')),
      $lessThan(variable('x'), variable('y'))[0]
    )(
      ...lessThanOps(
        'created less than',
        variableOps('x'),
        variableOps('y'),
        []
      )
    )
  })
})

describe('$greaterThan', () => {
  it('generates a GreaterThan for a pair of TreeNode inputs', () => {
    expect(
      $greaterThan(unit(variable('x').value), unit(variable('y').value))[0]
    ).toEqual({
      clade: Clades.binary, genus: Genera.inequalities, species: Species.greaterThan,
      left: unit(variable('x').value), right: unit(variable('y').value)
    })
  })
})

describe('greaterThan', () => {
  it('returns false for two ordered reals', () => {
    expectWriterTreeNode(
      greaterThan(real(1), real(2)),
      boolean(false)
    )(
      ...greaterThanOps(
        'real greater than',
        realOps('1'),
        realOps('2'),
        booleanOps('false')
      )
    )
  })

  it('returns true for two unordered reals', () => {
    expectWriterTreeNode(
      greaterThan(real(2), real(1)),
      boolean(true)
    )(
      ...greaterThanOps(
        'real greater than',
        realOps('2'),
        realOps('1'),
        booleanOps('true')
      )
    )
  })

  it('returns false for two equal inputs', () => {
    expectWriterTreeNode(
      greaterThan(real(1), real(1)),
      boolean(false)
    )(
      ...greaterThanOps(
        'real greater than',
        realOps('1'),
        realOps('1'),
        booleanOps('false')
      )
    )
  })

  it('returns false for two complexes ordered by length', () => {
    expectWriterTreeNode(
      greaterThan(complex([1, 1]), complex([5, 5])),
      boolean(false)
    )(
      ...greaterThanOps(
        'complex greater than',
        complexOps('1', '1'),
        complexOps('5', '5'),
        booleanOps('false')
      )
    )
  })

  it('returns true for two complexes not ordered by length', () => {
    expectWriterTreeNode(
      greaterThan(complex([5, 5]), complex([1, 1])),
      boolean(true)
    )(
      ...greaterThanOps(
        'complex greater than',
        complexOps('5', '5'),
        complexOps('1', '1'),
        booleanOps('true')
      )
    )
  })

  it('returns a GreaterThan for variable input', () => {
    expectWriterTreeNode(
      greaterThan(variable('x'), variable('y')),
      $greaterThan(variable('x'), variable('y'))[0]
    )(
      ...greaterThanOps(
        'created greater than',
        variableOps('x'),
        variableOps('y'),
        []
      )
    )
  })
})

describe('$lessThanEquals', () => {
  it('generates a LessThanEquals for a pair of TreeNode inputs', () => {
    expect(
      $lessThanEquals(unit(variable('x').value), unit(variable('y').value))[0]
    ).toEqual({
      clade: Clades.binary, genus: Genera.inequalities, species: Species.lessThanEquals,
      left: unit(variable('x').value), right: unit(variable('y').value)
    })
  })
})

describe('lessThanEquals', () => {
  it('returns true for two ordered reals', () => {
    expectWriterTreeNode(
      lessThanEquals(real(1), real(2)),
      boolean(true)
    )(
      ...lessThanEqualsOps(
        'real less than equals',
        realOps('1'),
        realOps('2'),
        booleanOps('true')
      )
    )
  })

  it('returns false for two unordered reals', () => {
    expectWriterTreeNode(
      lessThanEquals(real(2), real(1)),
      boolean(false)
    )(
      ...lessThanEqualsOps(
        'real less than equals',
        realOps('2'),
        realOps('1'),
        booleanOps('false')
      )
    )
  })

  it('returns true for two equal inputs', () => {
    expectWriterTreeNode(
      lessThanEquals(real(1), real(1)),
      boolean(true)
    )(
      ...lessThanEqualsOps(
        'real less than equals',
        realOps('1'),
        realOps('1'),
        booleanOps('true')
      )
    )
  })

  it('returns true for two complexes ordered by length', () => {
    expectWriterTreeNode(
      lessThanEquals(complex([1, 1]), complex([5, 5])),
      boolean(true)
    )(
      ...lessThanEqualsOps(
        'complex less than equals',
        complexOps('1', '1'),
        complexOps('5', '5'),
        booleanOps('true')
      )
    )
  })

  it('returns false for two complexes not ordered by length', () => {
    expectWriterTreeNode(
      lessThanEquals(complex([5, 5]), complex([1, 1])),
      boolean(false)
    )(
      ...lessThanEqualsOps(
        'complex less than equals',
        complexOps('5', '5'),
        complexOps('1', '1'),
        booleanOps('false')
      )
    )
  })

  it('returns a LessThanEquals for variable input', () => {
    expectWriterTreeNode(
      lessThanEquals(variable('x'), variable('y')),
      $lessThanEquals(variable('x'), variable('y'))[0]
    )(
      ...lessThanEqualsOps(
        'created less than equals',
        variableOps('x'),
        variableOps('y'),
        []
      )
    )
  })
})

describe('$greaterThanEquals', () => {
  it('generates a GreaterThanEquals for a pair of TreeNode inputs', () => {
    expect(
      $greaterThanEquals(unit(variable('x').value), unit(variable('y').value))[0]
    ).toEqual({
      clade: Clades.binary, genus: Genera.inequalities, species: Species.greaterThanEquals,
      left: unit(variable('x').value), right: unit(variable('y').value)
    })
  })
})

describe('greaterThanEquals', () => {
  it('returns false for two ordered reals', () => {
    expectWriterTreeNode(
      greaterThanEquals(real(1), real(2)),
      boolean(false)
    )(
      ...greaterThanEqualsOps(
        'real greater than equals',
        realOps('1'),
        realOps('2'),
        booleanOps('false')
      )
    )
  })

  it('returns true for two unordered reals', () => {
    expectWriterTreeNode(
      greaterThanEquals(real(2), real(1)),
      boolean(true)
    )(
      ...greaterThanEqualsOps(
        'real greater than equals',
        realOps('2'),
        realOps('1'),
        booleanOps('true')
      )
    )
  })

  it('returns true for two equal inputs', () => {
    expectWriterTreeNode(
      greaterThanEquals(real(1), real(1)),
      boolean(true)
    )(
      ...greaterThanEqualsOps(
        'real greater than equals',
        realOps('1'),
        realOps('1'),
        booleanOps('true')
      )
    )
  })

  it('returns false for two complexes ordered by length', () => {
    expectWriterTreeNode(
      greaterThanEquals(complex([1, 1]), complex([5, 5])),
      boolean(false)
    )(
      ...greaterThanEqualsOps(
        'complex greater than equals',
        complexOps('1', '1'),
        complexOps('5', '5'),
        booleanOps('false')
      )
    )
  })

  it('returns true for two complexes not ordered by length', () => {
    expectWriterTreeNode(
      greaterThanEquals(complex([5, 5]), complex([1, 1])),
      boolean(true)
    )(
      ...greaterThanEqualsOps(
        'complex greater than equals',
        complexOps('5', '5'),
        complexOps('1', '1'),
        booleanOps('true')
      )
    )
  })

  it('returns a GreaterThanEquals for variable input', () => {
    expectWriterTreeNode(
      greaterThanEquals(variable('x'), variable('y')),
      $greaterThanEquals(variable('x'), variable('y'))[0]
    )(
      ...greaterThanEqualsOps(
        'created greater than equals',
        variableOps('x'),
        variableOps('y'),
        []
      )
    )
  })
})
