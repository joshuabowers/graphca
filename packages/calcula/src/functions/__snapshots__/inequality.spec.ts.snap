// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`equals returns an Equals for variable input 1`] = `
[
  [
    "(x==y)",
    "identified equality",
  ],
  [
    "([x]==y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}==[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x=={y})",
    "processed right operand",
  ],
  [
    "(x==y)",
    "created equality",
  ],
]
`;

exports[`equals returns false for unequal things 1`] = `
[
  [
    "(1==2)",
    "identified equality",
  ],
  [
    "([1]==2)",
    "processing left operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "({1}==[2])",
    "processed left operand; processing right operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "(1=={2})",
    "processed right operand",
  ],
  [
    "(1==2)",
    "real equality",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`equals returns true for two equal booleans 1`] = `
[
  [
    "(false==false)",
    "identified equality",
  ],
  [
    "([false]==false)",
    "processing left operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "({false}==[false])",
    "processed left operand; processing right operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "(false=={false})",
    "processed right operand",
  ],
  [
    "(false==false)",
    "boolean equality",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`equals returns true for two equal complexes 1`] = `
[
  [
    "(1+1𝒊==1+1𝒊)",
    "identified equality",
  ],
  [
    "([1+1𝒊]==1+1𝒊)",
    "processing left operand",
  ],
  [
    "1+1𝒊",
    "created complex",
  ],
  [
    "({1+1𝒊}==[1+1𝒊])",
    "processed left operand; processing right operand",
  ],
  [
    "1+1𝒊",
    "created complex",
  ],
  [
    "(1+1𝒊=={1+1𝒊})",
    "processed right operand",
  ],
  [
    "(1+1𝒊==1+1𝒊)",
    "complex equality",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`equals returns true for two equal reals 1`] = `
[
  [
    "(1==1)",
    "identified equality",
  ],
  [
    "([1]==1)",
    "processing left operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "({1}==[1])",
    "processed left operand; processing right operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "(1=={1})",
    "processed right operand",
  ],
  [
    "(1==1)",
    "real equality",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`greaterThan returns a GreaterThan for variable input 1`] = `
[
  [
    "(x>y)",
    "identified greater than",
  ],
  [
    "([x]>y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}>[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x>{y})",
    "processed right operand",
  ],
  [
    "(x>y)",
    "created greater than",
  ],
]
`;

exports[`greaterThan returns false for two complexes ordered by length 1`] = `
[
  [
    "(1+1𝒊>5+5𝒊)",
    "identified greater than",
  ],
  [
    "([1+1𝒊]>5+5𝒊)",
    "processing left operand",
  ],
  [
    "1+1𝒊",
    "created complex",
  ],
  [
    "({1+1𝒊}>[5+5𝒊])",
    "processed left operand; processing right operand",
  ],
  [
    "5+5𝒊",
    "created complex",
  ],
  [
    "(1+1𝒊>{5+5𝒊})",
    "processed right operand",
  ],
  [
    "(1+1𝒊>5+5𝒊)",
    "complex greater than",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`greaterThan returns false for two equal inputs 1`] = `
[
  [
    "(1>1)",
    "identified greater than",
  ],
  [
    "([1]>1)",
    "processing left operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "({1}>[1])",
    "processed left operand; processing right operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "(1>{1})",
    "processed right operand",
  ],
  [
    "(1>1)",
    "real greater than",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`greaterThan returns false for two ordered reals 1`] = `
[
  [
    "(1>2)",
    "identified greater than",
  ],
  [
    "([1]>2)",
    "processing left operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "({1}>[2])",
    "processed left operand; processing right operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "(1>{2})",
    "processed right operand",
  ],
  [
    "(1>2)",
    "real greater than",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`greaterThan returns true for two complexes not ordered by length 1`] = `
[
  [
    "(5+5𝒊>1+1𝒊)",
    "identified greater than",
  ],
  [
    "([5+5𝒊]>1+1𝒊)",
    "processing left operand",
  ],
  [
    "5+5𝒊",
    "created complex",
  ],
  [
    "({5+5𝒊}>[1+1𝒊])",
    "processed left operand; processing right operand",
  ],
  [
    "1+1𝒊",
    "created complex",
  ],
  [
    "(5+5𝒊>{1+1𝒊})",
    "processed right operand",
  ],
  [
    "(5+5𝒊>1+1𝒊)",
    "complex greater than",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`greaterThan returns true for two unordered reals 1`] = `
[
  [
    "(2>1)",
    "identified greater than",
  ],
  [
    "([2]>1)",
    "processing left operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "({2}>[1])",
    "processed left operand; processing right operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "(2>{1})",
    "processed right operand",
  ],
  [
    "(2>1)",
    "real greater than",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`greaterThanEquals returns a GreaterThanEquals for variable input 1`] = `
[
  [
    "(x>=y)",
    "identified greater than equals",
  ],
  [
    "([x]>=y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}>=[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x>={y})",
    "processed right operand",
  ],
  [
    "(x>=y)",
    "created greater than equals",
  ],
]
`;

exports[`greaterThanEquals returns false for two complexes ordered by length 1`] = `
[
  [
    "(1+1𝒊>=5+5𝒊)",
    "identified greater than equals",
  ],
  [
    "([1+1𝒊]>=5+5𝒊)",
    "processing left operand",
  ],
  [
    "1+1𝒊",
    "created complex",
  ],
  [
    "({1+1𝒊}>=[5+5𝒊])",
    "processed left operand; processing right operand",
  ],
  [
    "5+5𝒊",
    "created complex",
  ],
  [
    "(1+1𝒊>={5+5𝒊})",
    "processed right operand",
  ],
  [
    "(1+1𝒊>=5+5𝒊)",
    "complex greater than equals",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`greaterThanEquals returns false for two ordered reals 1`] = `
[
  [
    "(1>=2)",
    "identified greater than equals",
  ],
  [
    "([1]>=2)",
    "processing left operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "({1}>=[2])",
    "processed left operand; processing right operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "(1>={2})",
    "processed right operand",
  ],
  [
    "(1>=2)",
    "real greater than equals",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`greaterThanEquals returns true for two complexes not ordered by length 1`] = `
[
  [
    "(5+5𝒊>=1+1𝒊)",
    "identified greater than equals",
  ],
  [
    "([5+5𝒊]>=1+1𝒊)",
    "processing left operand",
  ],
  [
    "5+5𝒊",
    "created complex",
  ],
  [
    "({5+5𝒊}>=[1+1𝒊])",
    "processed left operand; processing right operand",
  ],
  [
    "1+1𝒊",
    "created complex",
  ],
  [
    "(5+5𝒊>={1+1𝒊})",
    "processed right operand",
  ],
  [
    "(5+5𝒊>=1+1𝒊)",
    "complex greater than equals",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`greaterThanEquals returns true for two equal inputs 1`] = `
[
  [
    "(1>=1)",
    "identified greater than equals",
  ],
  [
    "([1]>=1)",
    "processing left operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "({1}>=[1])",
    "processed left operand; processing right operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "(1>={1})",
    "processed right operand",
  ],
  [
    "(1>=1)",
    "real greater than equals",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`greaterThanEquals returns true for two unordered reals 1`] = `
[
  [
    "(2>=1)",
    "identified greater than equals",
  ],
  [
    "([2]>=1)",
    "processing left operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "({2}>=[1])",
    "processed left operand; processing right operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "(2>={1})",
    "processed right operand",
  ],
  [
    "(2>=1)",
    "real greater than equals",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`lessThan returns a LessThan for variable input 1`] = `
[
  [
    "(x<y)",
    "identified less than",
  ],
  [
    "([x]<y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}<[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x<{y})",
    "processed right operand",
  ],
  [
    "(x<y)",
    "created less than",
  ],
]
`;

exports[`lessThan returns false for two complexes not ordered by length 1`] = `
[
  [
    "(5+5𝒊<1+1𝒊)",
    "identified less than",
  ],
  [
    "([5+5𝒊]<1+1𝒊)",
    "processing left operand",
  ],
  [
    "5+5𝒊",
    "created complex",
  ],
  [
    "({5+5𝒊}<[1+1𝒊])",
    "processed left operand; processing right operand",
  ],
  [
    "1+1𝒊",
    "created complex",
  ],
  [
    "(5+5𝒊<{1+1𝒊})",
    "processed right operand",
  ],
  [
    "(5+5𝒊<1+1𝒊)",
    "complex less than",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`lessThan returns false for two equal inputs 1`] = `
[
  [
    "(1<1)",
    "identified less than",
  ],
  [
    "([1]<1)",
    "processing left operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "({1}<[1])",
    "processed left operand; processing right operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "(1<{1})",
    "processed right operand",
  ],
  [
    "(1<1)",
    "real less than",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`lessThan returns false for two unordered reals 1`] = `
[
  [
    "(2<1)",
    "identified less than",
  ],
  [
    "([2]<1)",
    "processing left operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "({2}<[1])",
    "processed left operand; processing right operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "(2<{1})",
    "processed right operand",
  ],
  [
    "(2<1)",
    "real less than",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`lessThan returns true for two complexes ordered by length 1`] = `
[
  [
    "(1+1𝒊<5+5𝒊)",
    "identified less than",
  ],
  [
    "([1+1𝒊]<5+5𝒊)",
    "processing left operand",
  ],
  [
    "1+1𝒊",
    "created complex",
  ],
  [
    "({1+1𝒊}<[5+5𝒊])",
    "processed left operand; processing right operand",
  ],
  [
    "5+5𝒊",
    "created complex",
  ],
  [
    "(1+1𝒊<{5+5𝒊})",
    "processed right operand",
  ],
  [
    "(1+1𝒊<5+5𝒊)",
    "complex less than",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`lessThan returns true for two ordered reals 1`] = `
[
  [
    "(1<2)",
    "identified less than",
  ],
  [
    "([1]<2)",
    "processing left operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "({1}<[2])",
    "processed left operand; processing right operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "(1<{2})",
    "processed right operand",
  ],
  [
    "(1<2)",
    "real less than",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`lessThanEquals returns a LessThanEquals for variable input 1`] = `
[
  [
    "(x<=y)",
    "identified less than equals",
  ],
  [
    "([x]<=y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}<=[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x<={y})",
    "processed right operand",
  ],
  [
    "(x<=y)",
    "created less than equals",
  ],
]
`;

exports[`lessThanEquals returns false for two complexes not ordered by length 1`] = `
[
  [
    "(5+5𝒊<=1+1𝒊)",
    "identified less than equals",
  ],
  [
    "([5+5𝒊]<=1+1𝒊)",
    "processing left operand",
  ],
  [
    "5+5𝒊",
    "created complex",
  ],
  [
    "({5+5𝒊}<=[1+1𝒊])",
    "processed left operand; processing right operand",
  ],
  [
    "1+1𝒊",
    "created complex",
  ],
  [
    "(5+5𝒊<={1+1𝒊})",
    "processed right operand",
  ],
  [
    "(5+5𝒊<=1+1𝒊)",
    "complex less than equals",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`lessThanEquals returns false for two unordered reals 1`] = `
[
  [
    "(2<=1)",
    "identified less than equals",
  ],
  [
    "([2]<=1)",
    "processing left operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "({2}<=[1])",
    "processed left operand; processing right operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "(2<={1})",
    "processed right operand",
  ],
  [
    "(2<=1)",
    "real less than equals",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`lessThanEquals returns true for two complexes ordered by length 1`] = `
[
  [
    "(1+1𝒊<=5+5𝒊)",
    "identified less than equals",
  ],
  [
    "([1+1𝒊]<=5+5𝒊)",
    "processing left operand",
  ],
  [
    "1+1𝒊",
    "created complex",
  ],
  [
    "({1+1𝒊}<=[5+5𝒊])",
    "processed left operand; processing right operand",
  ],
  [
    "5+5𝒊",
    "created complex",
  ],
  [
    "(1+1𝒊<={5+5𝒊})",
    "processed right operand",
  ],
  [
    "(1+1𝒊<=5+5𝒊)",
    "complex less than equals",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`lessThanEquals returns true for two equal inputs 1`] = `
[
  [
    "(1<=1)",
    "identified less than equals",
  ],
  [
    "([1]<=1)",
    "processing left operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "({1}<=[1])",
    "processed left operand; processing right operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "(1<={1})",
    "processed right operand",
  ],
  [
    "(1<=1)",
    "real less than equals",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`lessThanEquals returns true for two ordered reals 1`] = `
[
  [
    "(1<=2)",
    "identified less than equals",
  ],
  [
    "([1]<=2)",
    "processing left operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "({1}<=[2])",
    "processed left operand; processing right operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "(1<={2})",
    "processed right operand",
  ],
  [
    "(1<=2)",
    "real less than equals",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`notEquals returns a NotEquals for variable input 1`] = `
[
  [
    "(x!=y)",
    "identified inequality",
  ],
  [
    "([x]!=y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}!=[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x!={y})",
    "processed right operand",
  ],
  [
    "(x!=y)",
    "created inequality",
  ],
]
`;

exports[`notEquals returns false for two equal booleans 1`] = `
[
  [
    "(false!=false)",
    "identified inequality",
  ],
  [
    "([false]!=false)",
    "processing left operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "({false}!=[false])",
    "processed left operand; processing right operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "(false!={false})",
    "processed right operand",
  ],
  [
    "(false!=false)",
    "boolean inequality",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`notEquals returns false for two equal complexes 1`] = `
[
  [
    "(1+1𝒊!=1+1𝒊)",
    "identified inequality",
  ],
  [
    "([1+1𝒊]!=1+1𝒊)",
    "processing left operand",
  ],
  [
    "1+1𝒊",
    "created complex",
  ],
  [
    "({1+1𝒊}!=[1+1𝒊])",
    "processed left operand; processing right operand",
  ],
  [
    "1+1𝒊",
    "created complex",
  ],
  [
    "(1+1𝒊!={1+1𝒊})",
    "processed right operand",
  ],
  [
    "(1+1𝒊!=1+1𝒊)",
    "complex inequality",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`notEquals returns false for two equal reals 1`] = `
[
  [
    "(1!=1)",
    "identified inequality",
  ],
  [
    "([1]!=1)",
    "processing left operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "({1}!=[1])",
    "processed left operand; processing right operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "(1!={1})",
    "processed right operand",
  ],
  [
    "(1!=1)",
    "real inequality",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`notEquals returns true for unequal things 1`] = `
[
  [
    "(1!=2)",
    "identified inequality",
  ],
  [
    "([1]!=2)",
    "processing left operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "({1}!=[2])",
    "processed left operand; processing right operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "(1!={2})",
    "processed right operand",
  ],
  [
    "(1!=2)",
    "real inequality",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;
